source/common/stats/stat_merger.h

```diff
@@ -0,0 +1,48 @@
+#pragma once
+
+#include <unordered_map>
+
+#include "envoy/stats/store.h"
+
+#include "common/protobuf/protobuf.h"
+
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Stats {
+
+// Responsible for the sensible merging of two instances of the same stat from two different
+// (typically hot restart parent+child) Envoy processes.
+class StatMerger {
+public:
+  StatMerger(Stats::Store& target_store);
+
+  enum class CombineLogic {
+    Accumulate,           // the default; the merged result is old+new.
+    OnlyImportWhenUnused, // import parent value only if child stat is undefined. (So, just once.)
+    NoImport, // ignore parent entirely; child stat is undefined until it sets its own value.
+  };
+
+  // Merge the values of stats_proto into stats_store. Counters are always straightforward
+  // addition, while gauges default to addition but have exceptions.
+  void mergeStats(const Protobuf::Map<std::string, uint64_t>& counters,
+                  const Protobuf::Map<std::string, uint64_t>& gauges);
+
+  // TODO(fredlas) add void verifyCombineLogicSpecified(absl::string_view gauge_name), to
+  // be called at gauge allocation, to ensure (with an ASSERT) that anyone adding a new stat
+  // will be forced to come across this code and explicitly specify combination logic.
+
+  // Looks up gauge_name in our nonstandard combine logic rules, and returns the result (nullopt
+  // meaning not nonstandard, implying Accumulate).
+  static absl::optional<StatMerger::CombineLogic> getCombineLogic(const std::string& gauge_name);
+
+private:
+  void mergeCounters(const Protobuf::Map<std::string, uint64_t>& counters);
+  void mergeGauges(const Protobuf::Map<std::string, uint64_t>& gauges);
+  std::unordered_map<std::string, uint64_t> parent_counter_values_;
```

Comment:
Are the keys here stat names? If so, rather than copying the entire string into the map, can you use instead use the name held inside the stat? See thread_local_store.cc for similar structures?

Also, how long does StatMerger typically live? During the overlap between child and parent? Can you specify that expectation in the class description?

Reply:
Note that even if this is short-lived, copying all the stat-names during the restart operation (say 1M stat names x 100 bytes per stat name) can change the high-water mark.

Reply:
Correct, it is only around while a hot restart is in progress. I will look into that.

Reply:
Done. Nice interface; that was surprisingly easy!